src/net.cpp

```diff
@@ -1143,6 +1143,30 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
     RandAddEvent((uint32_t)id);
 }
 
+bool CConnman::AddConnection(const std::string& address, ConnectionType conn_type)
+{
+    assert(conn_type == ConnectionType::OUTBOUND_FULL_RELAY || conn_type == ConnectionType::BLOCK_RELAY);
+
+    const int max_connections = conn_type == ConnectionType::OUTBOUND_FULL_RELAY ? m_max_outbound_full_relay : m_max_outbound_block_relay;
+
+    // Count existing connections
+    int existing_connections{0};
+    {
+        LOCK(cs_vNodes);
+        existing_connections = std::count_if(vNodes.begin(), vNodes.end(), [conn_type](CNode* node) { return node->m_conn_type == conn_type; });
+    }
```

Comment:
updated to `WITH_LOCK` syntax. 

we discussed this offline, but to document here for other reviewers- 

Since we hold the `cs_vNodes` lock for counting open connections but then release it before invoking `OpenNetworkConnection`, `AddConnection` and `ThreadOpenConnections` could both count, determine there's space for an extra, and attempt to open a block-relay-only. We could fix this by having a counter / boolean that indicates opening a connection is "in progress", but that seems unnecessary here because its a low likelihood and this is a test-only API.